### ISSUE TITLE
Importer for CodeChecker 

### ISSUE THREAD

**Issue by jay24rajput** [Ericsson/codechecker#3076]
**Is your feature request related to a problem? Please describe.**
We want to import the results generated from `CodeChecker cmd results` #3075 into another instance of the CodeChecker server. The formatted of the input file would be `json`. We will, for now, assume the users are on the same version of Codechecker and Kernel repository. 

**Rough Description of the scenario:**
- User A adds some comments to a specific run in the Codechecker server and marks some of the reports as false positive, confirm, etc. 
- User A runs `CodeChecker cmd results <run_name> -o json --details` and obtains a json file. 
- User A sends the thus obtained json to User B
- User B should now be able to import this json into his Codechecker server and see the findings and comments added by User A. 

We would be obliged if you could suggest an approach for implementing the functionality and redirect us to the appropriate files. @bulwahn

**Comment by gyorb** [comment]
@jay24rajput can we assume in this case that the user B can analyze the project and store the results with the source code?

**Comment by jay24rajput** [comment]
@gyorb In this case we assume that User B has already analyzed and stored the results into the server before receiving the json from User A.

Those are some really valid concerns. I think it would be appropriate that the contents of the database are overwritten. But, @bulwahn will be able to provide an appropriate answer to this scenario.

**Comment by bulwahn** [comment]
> @jay24rajput can we assume in this case that the user B can analyze the project and store the results with the source code?
>

Yes, we can assume that user A and B have ensured that they run the same tool on the same source code revision (or shared the tool output and both used the same report-generator and imported it similarly).

> Some issues which is worth to discuss is I think is to decide if the importer should overwrite the review/detection status of a report if there is a difference or not.

Agree. That should be an option to decide on such policies. I generally would like to extend the whole system to have true multi-user support where multiple users can have different review and detection status and some kind of moderation view and mode, but that is future work.

For now, let us have just some basic policies for checking/overwriting/reporting/aborting while importing.

> It here are already some user comment for a report the newly imported should only be added to that list or overwrite.
>

Agree. The importer should allow to follow different policies.
 
> Importing might lead to a state which can be confusing like there is a review comment (`codechecker_intentional`) in the source code but the in the imported data it different (`codechecker_false_positive`). In this case when the report is checked by the user the source file will contain a different review status compared to the state in the database. There should be at least some entry in the report history about this review status change done by an import action.
>

Agree.
 
> The current behavior is that the review comment in the source file will always overwrite the state in the database if the file was analyzed.
> 
> I think this is the same or connected to #2629

Agree. Although #2629 does not say much, we probably have a very similar feature in mind.

In the kernel community, we will have many people (probably for each tool around 5 to 10 on a pretty regular basis; but probably often also different additional people for each report finding) to make comments and each would have their own opinion on the state, and everyone wants to see the opinion of each other. Maybe CodeChecker is not built for this, though, but currently to me, it still looks like the best tool to grow in that direction.

**Comment by jay24rajput** [comment]
@csordasmarton @gyorb I think we can start off with the following approach considering all the assumptions stated above:

1. From the Json file received, filter out the comments(added in the exporter) and bughash
2. Overwrite the comments filtered in step 1 into the Codechecker database by using the bughash

What do you think about this? If it feels good, can you please redirect me to the files to be modified/created and some references for the same.

**Comment by jay24rajput** [comment]
@csordasmarton @gyorb I have developed command **import** for imported in `cmd.py` as stated above. Kindly guide me on how to actually implement the functionality and how to approach the same( eg where we can implement the logic for the same and if possible some examples of previous commands). 

Github branch for reference: https://github.com/jay24rajput/codechecker/tree/importer

**Comment by csordasmarton** [comment]
@jay24rajput Just to clarify things: do you want to export/import only comments (system/user comments and review statuses) or do you want to export/import full reports?

I am asking this because the second one is more complicated and we need to fetch and save a lot of extra data from the server (reports, comments, run, files, analyzer statistics etc.).

**Comment by jay24rajput** [comment]
@csordasmarton Yes as of now we only want to import comments and review status. The data will be the json file generated with the command: 
```sh
CodeChecker cmd results <run_name> -o json --details 2>/dev/null | python -m json.tool > <run_name>.json
```

What we would typically like to do as of now is, get all the comments/review status from the json(sent by User A) and override the comments/status of that particular run( which is already existing in User B's server)

**Comment by jay24rajput** [comment]
@csordasmarton I was have added the command `import` in `cmd.py`(https://github.com/Ericsson/codechecker/blob/master/web/client/codechecker_client/cmd/cmd.py) as:
```python3
importer = subcommands.add_parser(
        'import',
        formatter_class=arg.RawDescriptionDefaultHelpFormatter,
        description="Import the results into codechecker server",
        help="Import the analysis from a json file into"
        "codechecker for a given run"
    )
    __register_importer(importer)
    importer.set_defaults(func=cmd_line_client.handle_import)
    __add_common_arguments(importer)
```
I am defining a function `handle_import` in `command_line_client.py`(https://github.com/Ericsson/codechecker/blob/master/web/client/codechecker_client/cmd_line_client.py) where we will be:
- Checking if the run name is provided
- If not, take the run name from the json file provided
- For each report check if the comment message is same
- If the comment message is not the same:
    - Use removeComment(https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1773) to remove the comment from the database
    - Use addComment(https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1695) to add the comments mentioned in the json


Does this plan sound practical to you? If yes should we proceed with the same?

**Comment by bulwahn** [comment]
@jay24rajput I think the default should be to just add new messages and not delete existing ones. Other than that, this sounds good. Try to implement this and we can continue discussion on the refinement of this feature in your pull request.

**Comment by csordasmarton** [comment]
In this case maybe it's not the best approach. Lets assume that in your database you have 200.000 reports and you have only 1 comment for a bug. In this case you have to get all the results by using the `CodeChecker cmd results` command, process it on the client side, get the comments from it and then import it to the database. This is a little bit overkill for me.

Another problem can be if you want to import multiple comments to the server (e.g.: a hundred) and you use the `addComment` API function, it will work properly, but what if there is a network problem or the server is not reachable. It will import some comments but not all of them. We have to import data in one transaction. 

So I am thinking on the following solution. Lets create two subcommands:
- `CodeChecker cmd export export.json`: this will export the data from the server to a file. For now it will export only comments and review statuses but later we can extend it with more information if needed.
- `CodeChecker cmd import export.json`: this will import the data from the file to the server.

For this we have to do the following:
- Register two subcommands (`export`/`import`) similar to this one:
https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/client/codechecker_client/cmd/cmd.py#L1343-L1351
- Extend the thrift API (this file: https://github.com/Ericsson/codechecker/blob/master/web/api/report_server.thrift) with the following things:
  - Introduce a new type to export/import data:
    ```thrift
    struct ExportData {
      1: map<string, CommentDataList> comments,   // Map comments to report hashes.
      2: map<string, ReviewData>      reviewData, // Map review data to report hashes.
    }
    ```
  - Introduce new API functions:
    ```thrift
      // Export data from the server
      // PERMISSION: PRODUCT_ACCESS
      ExportData exportData()
                            throws (1: codechecker_api_shared.RequestFailed requestError),
    
      // Import data from the server.
      // PERMISSION: PRODUCT_ADMIN
      bool importData(1: ExportData export)
                      throws (1: codechecker_api_shared.RequestFailed requestError),
    ```
    My recommendation here: anybody can export data from the server but to import data to the server the user needs admin rights.
- Implement the new API functions in this file similar to this:
https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L2496-L2502

We can implement some logic here:
- when importing review statuses and if the `date` field is the same which can be found in the database for the report hash, we don't touch the data in the database because it is up to date. Otherwise we will change the review status and we can create new system comments, so in the comments sidebar we will see who changed the review status, when and from what.
- If a comment already exists in the database (we can check if the `id` field exists in the database and the message is the same) the server will not add it again to the database otherwise it will create a new comment or update the existing one (in this case we can also create a system comment).

You can find more information how to add/update comments here:
https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1695
https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1722
And you can reuse this API function to set/update the review status:
https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1551

What do you think about this solution @jay24rajput, @bulwahn ?

**Comment by bulwahn** [comment]
> In this case maybe it's not the best approach. Lets assume that in your database you have 200.000 reports and you have only 1 comment for a bug. In this case you have to get all the results by using the `CodeChecker cmd results` command, process it on the client side, get the comments from it and then import it to the database. This is a little bit overkill for me.
>

Maybe I just misunderstood @jay24rajput's proposal; I was thinking that only for those reports where a comment would be added, the command checks if the report is there (just as sanity check) because we basically assume that the reports between user A and user B on the same run are identical (otherwise, we have a complete different problem and this import feature cannot work at all).
 
> Another problem can be if you want to import multiple comments to the server (e.g.: a hundred) and you use the `addComment` API function, it will work properly, but what if there is a network problem or the server is not reachable. It will import some comments but not all of them. We have to import data in one transaction.
> 
> So I am thinking on the following solution. Lets create two subcommands:
> 
> * `CodeChecker cmd export export.json`: this will export the data from the server to a file. For now it will export only comments and review statuses but later we can extend it with more information if needed.

This functionality already exists, right? I mean: we can already obtain a json file with that information.

> * `CodeChecker cmd import export.json`: this will import the data from the file to the server.
> 
> For this we have to do the following:
> 
> * Register two subcommands (`export`/`import`) similar to this one:
>   https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/client/codechecker_client/cmd/cmd.py#L1343-L1351
> * Extend the thrift API (this file: https://github.com/Ericsson/codechecker/blob/master/web/api/report_server.thrift) with the following things:
>   
>   * Introduce a new type to export/import data:
>     ```thrift
>     struct ExportData {
>       1: map<string, CommentDataList> comments,   // Map comments to report hashes.
>       2: map<string, ReviewData>      reviewData, // Map review data to report hashes.
>     }
>     ```
>   * Introduce new API functions:
>     ```thrift
>       // Export data from the server
>       // PERMISSION: PRODUCT_ACCESS
>       ExportData exportData()
>                             throws (1: codechecker_api_shared.RequestFailed requestError),
>     
>       // Import data from the server.
>       // PERMISSION: PRODUCT_ADMIN
>       bool importData(1: ExportData export)
>                       throws (1: codechecker_api_shared.RequestFailed requestError),
>     ```
>     
>     
>     My recommendation here: anybody can export data from the server but to import data to the server the user needs admin rights.

I guess for now that permission scheme is okay.

> * Implement the new API functions in this file similar to this:
>   https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L2496-L2502
> 
> We can implement some logic here:
> 
> * when importing review statuses and if the `date` field is the same which can be found in the database for the report hash, we don't touch the data in the database because it is up to date. Otherwise we will change the review status and we can create new system comments, so in the comments sidebar we will see who changed the review status, when and from what.
> * If a comment already exists in the database (we can check if the `id` field exists in the database and the message is the same) the server will not add it again to the database otherwise it will create a new comment or update the existing one (in this case we can also create a system comment).
> 
> You can find more information how to add/update comments here:
> https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1695
> 
> 
> https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1722
> 
> 
> And you can reuse this API function to set/update the review status:
> https://github.com/Ericsson/codechecker/blob/792bd8150210c6cdd2f2ac728ed75ce31d77972c/web/server/codechecker_server/api/report_server.py#L1551
> 
> What do you think about this solution @jay24rajput, @bulwahn ?

I think the solution would work and allows us to try out some basic importing. We will see later how we need refine that with multiple users exchanging such .json files and exporting and importing them among each other.

**Comment by csordasmarton** [comment]
Every report has a unique ID (reportId) and a unique hash (report hash) field. Multiple reports may have the same report hash (for example if you store the same results to different runs or if a bug can be found through different file paths). For this reason the comments and review statuses are connected to report hashes and not for report ID's.

The problem with the `CodeChecker cmd results` command that it will return with all reports (json output is supported), but it will return not only comments but a lot more information. And if a comment is connected to multiple reports we will return with the same comment multiple times.

For this reason my recommendation was to create separate commands which will export and import things what are necessary for us. This way the output files will be smaller and the feature will be faster too. And later if we need more information it will be easier to extend these export/import API functions.

**Comment by jay24rajput** [comment]
@csordasmarton I have implemented the handler function for command for the `cmd export` command. Even added `struct` and API functions for the same. I think you will need to update the npm and pypi packages for my repository. Because the newly implemented function `getExportData` is not getting recognized by codechecker. Following is the error stack:
```
Traceback (most recent call last):
  File "/home/jay/codechecker/build/CodeChecker/cc_bin/CodeChecker.py", line 141, in main
    sys.exit(args.func(args))
  File "/home/jay/codechecker/build/CodeChecker/lib/python3/codechecker_client/cmd_line_client.py", line 1662, in handle_export
    report = client.getExportData(run.runid)
AttributeError: 'ThriftResultsHelper' object has no attribute 'getExportData'
```
I am attaching the link to my GitHub branch. Please tell me if anything else needs to be done to resolve the error

Github branch: https://github.com/jay24rajput/codechecker/tree/cmd_exporter

**Comment by csordasmarton** [comment]
@jay24rajput I have created the new npm/pypi packages (`6.37.0-dev2`). Just run the following command in the root of your repository to update the versions in your repository: `cd web/api && ./change-api-version.sh 6.37.0-dev2`

**Comment by jay24rajput** [comment]
@csordasmarton it is still not recognizing the function despite the error.  I also tried
```sh
make clean_package && make package
```

I tried the command in the codechecker root. I also tried it using virtual environment and even without the venv. I can see that the dependencies have been updated in the `requirement.txt` and other given files. 

Can you suggest any other commands required for making it work?

**Comment by csordasmarton** [comment]
You have to register the API function in this file if you want to call it from the command line: https://github.com/Ericsson/codechecker/blob/master/web/client/codechecker_client/helpers/results.py

```py
    @ThriftClientCall
    def exportData(self, run_id):
        pass
```

By the way the `exportData` API function doesn't have a run id parameter now:
https://github.com/jay24rajput/codechecker/blob/83533378b06a089f73350d6d48375744799bafa2/web/api/report_server.thrift#L735-L736
but in your code you give parameter to this api function, so it will raise an exception.

Also you didn't create an API function to import data in this thrift file so it will not be available in your code.

Please modify it to the following:
```thrift
  // Export data from the server
  // PERMISSION: PRODUCT_ACCESS
  ExportData exportData(1: i64 runId)
                        throws (1: codechecker_api_shared.RequestFailed requestError),

  // Import data from the server.
  // PERMISSION: PRODUCT_ADMIN
  bool importData(1: ExportData export)
                  throws (1: codechecker_api_shared.RequestFailed requestError),
```

I have created new packages based on this changes: `6.37.0-dev3`.

**Comment by jay24rajput** [comment]
@csordasmarton So I am getting an error with the `make ` command. This is what I did:
1. Reset the working branch with upstream/cmd_exporter:
`git reset --hard upstream/cmd_exporter`
2. Rebase with origin master: 
```sh
git fetch origin
git rebase origin/master
```
3. Made the changes suggested by you to `report_server.thrift` and `helpers.py` 
4. Updated repository to new packages: `cd web/api && ./change-api-version.sh 6.37.0-dev3`
5. Clean and make package: `make clean && make package`

After running the make package command I am getting the following error:
```sh
ERROR in ./node_modules/codechecker-api/lib/codeCheckerDBAccess.js 10466:58
Module parse failed: Unexpected keyword 'export' (10466:58)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|   return callback('exportData failed: unknown result');
| };
> codeCheckerDBAccessClient.prototype.importData = function(export, callback) {
|   this._seqid = this.new_seqid();
|   if (callback === undefined) {
 @ ./src/services/api/cc.service.js 61:0-42 90:51-64
 @ ./src/services/api/index.js
 @ ./src/main.js
Child HtmlWebpackCompiler:
     1 asset
    Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
    [MFLM] ./node_modules/html-webpack-plugin/lib/loader.js!./src/index.html 903 bytes {0} [built]
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! codechecker@1.0.0 build: `rimraf dist && cross-env NODE_ENV=production webpack --config config/webpack.prod.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the codechecker@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jay/.npm/_logs/2020-12-20T12_49_09_179Z-debug.log
make[1]: *** [Makefile:73: /home/jay/codechecker/web/server//vue-cli/dist] Error 2
make[1]: Leaving directory '/home/jay/codechecker/web'
make: *** [Makefile:100: package] Error 2
```

Am I doing something wrong here?

**Comment by csordasmarton** [comment]
Sorry, my bad  :confused:  We declared the importData function with one parameter which name is `export`. When thrift generated javascript functions from it, the parameter name will be `export` which is a reserved keyword. So we need to change the thrift file to this:
```thrift
bool importData(1: ExportData exportData)
                throws (1: codechecker_api_shared.RequestFailed requestError),
```
I have also generated a new package from it: `6.37.0-dev4`. Please try it out, now it should work.

Next time please always update your branch with your changes, so it will be easier to debug this for me :blush:

**Comment by jay24rajput** [comment]
> Sorry, my bad  We declared the importData function with one parameter which name is `export`. When thrift generated javascript functions from it, the parameter name will be `export` which is a reserved keyword. So we need to change the thrift file to this:
> 
> ```thrift
> bool importData(1: ExportData exportData)
>                 throws (1: codechecker_api_shared.RequestFailed requestError),
> ```
> 
> I have also generated a new package from it: `6.37.0-dev4`. Please try it out, now it should work.
> 
Yes this worked! Thanks for your help 😄 

> Next time please always update your branch with your changes, so it will be easier to debug this for me

Yes, I will keep that in mind from next time 😅

**Comment by jay24rajput** [comment]
@csordasmarton @bulwahn 
I have implemented the logic for exporter command bt when the command is called I get the following error:
```sh
[ERROR 2020-12-20 21:59] - Thrift invalid data error.
[ERROR 2020-12-20 21:59] - exportData
[ERROR 2020-12-20 21:59] - (1,)
[ERROR 2020-12-20 21:59] - {}
[ERROR 2020-12-20 21:59] - Request failed.
Traceback (most recent call last):
  File "/home/jay/codechecker/build/CodeChecker/lib/python3/codechecker_client/thrift_call.py", line 37, in wrapper
    return func(*args, **kwargs)
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/codechecker_api/codeCheckerDBAccess_v6/codeCheckerDBAccess.py", line 2048, in exportData
    return self.recv_exportData()
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/codechecker_api/codeCheckerDBAccess_v6/codeCheckerDBAccess.py", line 2067, in recv_exportData
    result.read(iprot)
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/codechecker_api/codeCheckerDBAccess_v6/codeCheckerDBAccess.py", line 10826, in read
    self.success.read(iprot)
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/codechecker_api/codeCheckerDBAccess_v6/ttypes.py", line 3244, in read
    _val214.read(iprot)
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/codechecker_api/codeCheckerDBAccess_v6/ttypes.py", line 1637, in read
    iprot.readStructBegin()
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/thrift/protocol/TJSONProtocol.py", line 430, in readStructBegin
    self.readJSONObjectStart()
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/thrift/protocol/TJSONProtocol.py", line 395, in readJSONObjectStart
    self.context.read()
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/thrift/protocol/TJSONProtocol.py", line 144, in read
    self.doIO(self.protocol.readJSONSyntaxChar)
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/thrift/protocol/TJSONProtocol.py", line 137, in doIO
    function(COLON if self.colon else COMMA)
  File "/home/jay/codechecker/venv/lib/python3.8/site-packages/thrift/protocol/TJSONProtocol.py", line 255, in readJSONSyntaxChar
    raise TProtocolException(TProtocolException.INVALID_DATA,
thrift.protocol.TProtocol.TProtocolException: Unexpected character: b'['
```

The error occurs while calling the function on this line: https://github.com/jay24rajput/codechecker/blob/0f9e613f41d06869afbdd9964ae603ecba5270c6/web/client/codechecker_client/cmd_line_client.py#L1662

Could you please guide me as to how this error could be solved? Also, have I implemented the proper return type and logic here:https://github.com/jay24rajput/codechecker/blob/0f9e613f41d06869afbdd9964ae603ecba5270c6/web/server/codechecker_server/api/report_server.py#L3298

When i tried to replace the return statement with print, I was getting the appropriate results

**Comment by csordasmarton** [comment]
The problem is in this place: https://github.com/jay24rajput/codechecker/blob/0f9e613f41d06869afbdd9964ae603ecba5270c6/web/server/codechecker_server/api/report_server.py#L3317

For each bug hash there is only one review data object:
```thrift
map<string, ReviewData>      reviewData,
```
See: https://github.com/jay24rajput/codechecker/blob/0f9e613f41d06869afbdd9964ae603ecba5270c6/web/api/report_server.thrift#L367

The problem is that you are trying to return with a *list* of review data object for each report hash.

**Comment by jay24rajput** [comment]
Thanks a ton @csordasmarton! The issue is resolved. We can obtain a proper formatted json. I will soon open after cleaning and formatting the code 🙌🏻
